### PR TITLE
mysql: refine decimal

### PR DESCRIPTION
**Pull request by coocood**
Prepare for replace old decimal with new decimal.

The changes includes:

1. encode 'precision' and 'frac' at the beginning in 'ToBin', so `FromBin` can decode without type information
2. replace errcode with error.
3. remove roundMode, always use halfUP
4. extract `digitsToWords` method.

### CODE REVIEW

File: mysql/error.go [hunk 1]
```diff
@@ -22,6 +22,11 @@ import (
 var (
 	ErrBadConn       = errors.New("connection was bad")
 	ErrMalformPacket = errors.New("Malform packet error")
+
+	ErrTruncated = errors.New("truncated")
```
Comment (shenli) [inline mysql/error.go]: Why not use terror?

File: mysql/mydecimal.go [hunk 2]
```diff
@@ -695,44 +680,30 @@ func (d *MyDecimal) doMiniRightShift(shift, beg, end int) {
 //
 // RETURN VALUE
 //  eDecOK/eDecTruncated
-func (d *MyDecimal) Round(to *MyDecimal, frac int, mode roundMode) (errcode int) {
+func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
```
Comment (shenli) [inline mysql/mydecimal.go]: AFAIK MySQL has half even rounding. https://dev.mysql.com/doc/refman/5.7/en/precision-math-rounding.html

File: mysql/mydecimal.go [hunk 3]
```diff
@@ -1173,12 +1153,19 @@ func (d *MyDecimal) ToBin(precision, frac int) ([]byte, int) {
...
 
 // FromBin Restores decimal from its binary fixed-length representation.
-func (d *MyDecimal) FromBin(bin []byte, precision, frac int) int {
-	var errcode = eDecOK
+func (d *MyDecimal) FromBin(bin []byte) error {
+	if len(bin) < 3 {
+		*d = zeroMyDecimal
+		return ErrBadNumber
+	}
+	precision := int(bin[0])
```
Comment (shenli) [inline mysql/mydecimal.go]: Why not use int8?

**Comment by shenli** [comment]
Rest LGTM

File: mysql/error.go (hunk 1 above)
Comment (coocood) [inline mysql/error.go]: terror depends on 'mysql' package.

File: mysql/mydecimal.go (hunk 2 above)
Comment (coocood) [inline mysql/mydecimal.go]: half even is used for float.
decimal only use half up.

File: mysql/mydecimal.go (hunk 3 above)
Comment (coocood) [inline mysql/mydecimal.go]: Use 'int' during calculation, cast to 'int8' when set the result value, this way can avoid overflow issue.

**Comment by coocood** [comment]
@tiancaiamao PTAL

**Comment by shenli** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**